{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "# from operator import itemgetter #, attrgetter\n",
    "import copy, os\n",
    "from math import pow\n",
    "#import random, sys\n",
    "import numpy as np\n",
    "import re\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "%matplotlib inline "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Basic(object):\n",
    "    def __init__(self):\n",
    "        self.log202 = np.log10(20) / np.log10(2)\n",
    "        self.log2 = np.log10(2)\n",
    "        self.ln2 = np.log(2)\n",
    "        self.inv_ln2 = 1./self.ln2\n",
    "        self.ln2_2 = self.ln2 ** 2\n",
    "        self.convShannonTsallis = 1.417199516  # =1/LN(2)\n",
    "    \n",
    "    def getDnaNucleotides(self):\n",
    "        \n",
    "        return ['A','T','G','C']\n",
    "    \n",
    "    def getDnaNucleotideString(self):\n",
    "        return 'ATGC'\n",
    "    \n",
    "    def getRnaNucleotides(self):\n",
    "        return ['A','U','G','C']\n",
    "    \n",
    "    def getrnaNucleotideString(self):\n",
    "        return 'AUGC'\n",
    "\n",
    "    # in Hydropathy index[95] http://en.wikipedia.org/wiki/Amino_acid\n",
    "    def getSeqAA(self):\n",
    "        return ['A', 'M', 'C', 'F', 'L', 'V', 'I', 'G', 'T', 'S', 'W', 'Y', 'P', 'H', 'N', 'D', 'E', 'Q', 'K', 'R']\n",
    "    \n",
    "    def getStringAA(self):\n",
    "        return 'AMCFLVIGTSWYPHNDEQKR'\n",
    "    \n",
    "    def getAaPos(self, string, aa):\n",
    "        return string.find(aa)\n",
    "    \n",
    "\n",
    "    def tira_aspas(self, stri):\n",
    "        if (stri == None):\n",
    "            return 'Null'\n",
    "        \n",
    "        try:\n",
    "            stri = stri.replace(\"'\",\"\")\n",
    "            stri = stri.replace('\"',\"'\")\n",
    "        except:\n",
    "            stri = str(stri)\n",
    "            stri = stri.replace(\"'\",\"\")\n",
    "            stri = stri.replace('\"',\"'\")\n",
    "            stri = stri.replace('[',\"\")\n",
    "            stri = stri.replace(']',\"\")\n",
    "            \n",
    "        return stri \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class MySequence:\n",
    "    def __init__(self, prjName = \"Human_papillomavirus\", root = \"fasta/\"):\n",
    "        self.clearSequences()\n",
    "        self.NucleotideAlphabet = [\"A\",\"C\",\"G\",\"T\"]\n",
    "        \n",
    "        self.prjName = prjName\n",
    "        self.root    = root\n",
    "        \n",
    "        self.isProtein = False\n",
    "        \n",
    "        self.clearSequences()\n",
    "        \n",
    "        \n",
    "    def read(self, typeSeq = \"L2\", pattern = None, numOfRecords=None, showmessage=False):\n",
    "\n",
    "        self.typeSeq = typeSeq\n",
    "        self.pattern = pattern\n",
    "        self.showmessage = showmessage\n",
    "     \n",
    "        if pattern == None:\n",
    "            self.fileName    = \"%s%s_%s.fasta\"%(       self.root, self.prjName, typeSeq)\n",
    "            self.fileNameAln = \"%s%s_%s_aligned.fasta\"%(self.root, self.prjName, typeSeq)\n",
    "        else:\n",
    "            self.fileName    = \"%s%s_%s_type_%d.fasta\"%(        self.root, self.prjName, typeSeq, pattern)\n",
    "            self.fileNameAln = \"%s%s_%s_type_%d_aligned.fasta\"%(self.root, self.prjName, typeSeq, pattern)\n",
    "            \n",
    "        # self.basic = Basic()\n",
    "        ret = self.readMySequence(self.fileName, numOfRecords, showmessage)\n",
    "\n",
    "    def read_aligned(self, typeSeq = \"L2\", pattern = None, suffix = None, numOfRecords=None, showmessage=False):\n",
    "\n",
    "        self.typeSeq = typeSeq\n",
    "        self.pattern = pattern\n",
    "        self.showmessage = showmessage\n",
    "\n",
    "        if pattern == None:\n",
    "            if suffix == None:\n",
    "                self.fileNameAln = \"%s%s_%s_aligned.fasta\"%(self.root, self.prjName, typeSeq)\n",
    "            else:\n",
    "                self.fileNameAln = \"%s%s_%s_%s_aligned.fasta\"%(self.root, self.prjName, typeSeq, sufix)\n",
    "        else:\n",
    "            if suffix == None:\n",
    "                self.fileNameAln = \"%s%s_%s_type_%d_aligned.fasta\"%(self.root, self.prjName, typeSeq, pattern)\n",
    "            else:\n",
    "                self.fileNameAln = \"%s%s_%s_type_%d_%s_aligned.fasta\"%(self.root, self.prjName, typeSeq, pattern, suffix)\n",
    "            \n",
    "        # self.basic = Basic()\n",
    "        ret = self.readMySequence(self.fileNameAln, numOfRecords, showmessage)\n",
    "\n",
    "    def clearSequences(self):\n",
    "        self.seqs = []\n",
    "        self.seq_records = []\n",
    "        \n",
    "        self.nrow = 0\n",
    "        self.ncol = 0\n",
    "\n",
    "        \n",
    "    def readMySequence(self, fileName, numOfRecords=None, showmessage=False):\n",
    "        self.clearSequences()\n",
    "        \n",
    "        if not os.path.isfile(fileName):\n",
    "            print(\"Could not find '%s'\"%fileName)\n",
    "            return False\n",
    "        \n",
    "        print(\"Reading ... '%s'\"%fileName)\n",
    "        \n",
    "        if numOfRecords == None:\n",
    "            try:\n",
    "                for seq_record in SeqIO.parse(fileName, \"fasta\"):\n",
    "                    self.seq_records.append(seq_record)\n",
    "                    self.seqs.append( list(str(seq_record.seq)) )\n",
    "\n",
    "            except ValueError:\n",
    "                print(\"Exception %s: error reading file: '%s'\"%(str(ValueError), fileName) )\n",
    "                return False\n",
    "            \n",
    "        else:\n",
    "            try:\n",
    "                count = 0\n",
    "                for seq_record in SeqIO.parse(fileName, \"fasta\"):\n",
    "                    self.seq_records.append(seq_record)\n",
    "                    self.seqs.append( list(str(seq_record.seq)) )\n",
    "                    count += 1\n",
    "                    if count == numOfRecords: break\n",
    "                    \n",
    "            except ValueError:\n",
    "                print(\"Exception %s: error reading file: '%s'\"%(str(ValueError), fileName) )\n",
    "                return False\n",
    "       \n",
    "\n",
    "        if (showmessage):\n",
    "            print(\"Read %d seqs from '%s'\" %(len(self.seq_records), fileName) )\n",
    "            \n",
    "        self.seqs = np.array(self.seqs)\n",
    "        \n",
    "        self.nrow = self.seqs.shape[0]\n",
    "        \n",
    "        try:\n",
    "            self.ncol = self.seqs.shape[1]\n",
    "\n",
    "        except:\n",
    "            self.ncol = 1\n",
    "            \n",
    "            print(\"Problems: fasta is not aligned. Impossible convert into matrix\")\n",
    "            return False\n",
    "        \n",
    "        return True\n",
    "    \n",
    "\n",
    "    def search_term(self, term, verbose = False):\n",
    "        records = []; count = 0\n",
    "        \n",
    "        for record in self.seq_records:\n",
    "            x = re.search(term, record.description)\n",
    "            if x != None:\n",
    "                count += 1\n",
    "                if verbose: print(count, \"\\t\", record.description)      \n",
    "                records.append(record)\n",
    "                \n",
    "        return records\n",
    "    \n",
    "    def write_seqs(self, seq_records, fname):\n",
    "        try:\n",
    "            SeqIO.write(seq_records, fname, \"fasta\")\n",
    "            print(\"%d records saved at %s\"%(len(seq_records), fname))\n",
    "\n",
    "        except ValueError:\n",
    "            print(\"Could not write - error %s; file: '%s'\"%(str(ValueError), fname) )\n",
    "            return False\n",
    "\n",
    "\n",
    "        return True\n",
    "        \n",
    "    def allign_with_muscle(self, in_file, out_file, muscleApp = r\"/media/flalix/disk2t/seaview/muscle\"):\n",
    "        if not os.path.isfile(muscleApp):\n",
    "            print(\"Could not find muscle at '%s'\", muscleApp)\n",
    "            return False\n",
    "\n",
    "        if not os.path.isfile(out_file):\n",
    "            muscle_cline = MuscleCommandline(muscleApp, input=in_file, out=out_file)\n",
    "            print(muscle_cline)\n",
    "            os.system(\"%s\"%(muscle_cline))\n",
    "        else:\n",
    "            print(\"File exists: '%s'\"%out_file)\n",
    "            \n",
    "        return True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read HLA type A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "prjName = \"HLA\"; root = \"fasta/\"\n",
    "mseq1 = MySequence(prjName = \"HLA\", root = \"fasta/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading ... 'fasta/HLA_A.fasta'\n",
      "Read 40531 seqs from 'fasta/HLA_A.fasta'\n",
      "Problems: fasta is not aligned. Impossible convert into matrix\n"
     ]
    }
   ],
   "source": [
    "mseq1.read(typeSeq = \"A\", pattern = None, numOfRecords=None, showmessage=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------------\n",
      "Processing the record lcl|Z46633.1_cds_CAA86602.1_1:\n",
      "Its description is: \n",
      "lcl|Z46633.1_cds_CAA86602.1_1 [gene=hla-A] [db_xref=GDB:119310,GOA:P01892,HGNC:4931,InterPro:IPR001039,InterPro:IPR003006,InterPro:IPR003597,InterPro:IPR007110,InterPro:IPR010579,InterPro:IPR011161,InterPro:IPR011162,InterPro:IPR013783,PDB:1AKJ,PDB:1AO7,PDB:1AQD,PDB:1B0G,PDB:1B0R,PDB:1BD2,PDB:1DUY,PDB:1DUZ,PDB:1EEY,PDB:1EEZ,PDB:1HHG,PDB:1HHH,PDB:1HHI,PDB:1HHJ,PDB:1HHK,PDB:1HLA,PDB:1I1F,PDB:1I1Y,PDB:1I4F,PDB:1I7R,PDB:1I7T,PDB:1I7U,PDB:1IM3,PDB:1JF1,PDB:1JHT,PDB:1LP9,PDB:1OGA,PDB:1P7Q,PDB:1QEW,PDB:1QR1,PDB:1QRN,PDB:1QSE,PDB:1QSF,PDB:1S8D,PDB:1S9W,PDB:1S9X,PDB:1S9Y,PDB:1T1W,PDB:1T1X,PDB:1T1Y,PDB:1T1Z,PDB:1T20,PDB:1T21,PDB:1T22,PDB:1TVB,PDB:1TVH,PDB:1UR7,PDB:2AV1,PDB:2AV7,PDB:2BNQ,PDB:2BNR,PDB:2C7U,PDB:2CLR,PDB:2F53,PDB:2F54,PDB:2GIT,PDB:2GJ6,PDB:2GT9,PDB:2GTW,PDB:2GTZ,PDB:2GUO,PDB:2J8U,PDB:2JCC,PDB:2P5E,PDB:2P5W,PDB:2PYE,PDB:2UWE,PDB:2V2W,PDB:2V2X,PDB:2VLJ,PDB:2VLK,PDB:2VLL,PDB:2VLR,PDB:2X4N,PDB:2X4O,PDB:2X4P,PDB:2X4Q,PDB:2X4R,PDB:2X4S,PDB:2X4T,PDB:2X4U,PDB:3BGM,PDB:3BH8,PDB:3BH9,PDB:3BHB,PDB:3D25,PDB:3D39,PDB:3D3V,PDB:3FQN,PDB:3FQR,PDB:3FQT,PDB:3FQU,PDB:3FQW,PDB:3FQX,PDB:3FT2,PDB:3FT3,PDB:3FT4,PDB:3GIV,PDB:3GJF,PDB:3GSN,PDB:3GSO,PDB:3GSQ,PDB:3GSR,PDB:3GSU,PDB:3GSV,PDB:3GSW,PDB:3GSX,PDB:3H7B,PDB:3H9H,PDB:3H9S,PDB:3HAE,PDB:3HG1,PDB:3HLA,PDB:3HPJ,PDB:3I6G,PDB:3I6K,PDB:3IXA,PDB:3KLA,PDB:3MGO,PDB:3MGT,PDB:3MYJ,UniProtKB/Swiss-Prot:P01892] [protein=lymphocyte antigen] [protein_id=CAA86602.1] [location=1..1098] [gbkey=CDS]\n",
      "Its sequence contains 1098 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|HE978329.1_cds_CCK73125.1_1:\n",
      "Its description is: \n",
      "lcl|HE978329.1_cds_CCK73125.1_1 [gene=HLA-A] [frame=3] [protein_id=CCK73125.1] [location=join(<1..270,371..646,747..>1022)] [gbkey=CDS]\n",
      "Its sequence contains 820 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|HE978328.1_cds_CCK73124.1_1:\n",
      "Its description is: \n",
      "lcl|HE978328.1_cds_CCK73124.1_1 [gene=HLA-A] [frame=3] [protein_id=CCK73124.1] [location=join(<1..270,371..646,747..>1022)] [gbkey=CDS]\n",
      "Its sequence contains 820 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|D38525.1_cds_BAA07530.1_1:\n",
      "Its description is: \n",
      "lcl|D38525.1_cds_BAA07530.1_1 [gene=HLA-A] [protein_id=BAA07530.1] [location=1..843] [gbkey=CDS]\n",
      "Its sequence contains 843 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|KU563153.1_cds_ANW09655.1_1:\n",
      "Its description is: \n",
      "lcl|KU563153.1_cds_ANW09655.1_1 [gene=HLA-A] [protein=HLA-A] [frame=3] [protein_id=ANW09655.1] [location=join(<1..270,371..646,747..>1022)] [gbkey=CDS]\n",
      "Its sequence contains 820 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|AH004889.2_cds_AAB39110.1_1:\n",
      "Its description is: \n",
      "lcl|AH004889.2_cds_AAB39110.1_1 [gene=HLA-A] [protein=HLA-A] [frame=3] [protein_id=AAB39110.1] [location=join(<1..270,371..>646)] [gbkey=CDS]\n",
      "Its sequence contains 544 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|KC296460.1_cds_AGH58093.1_1:\n",
      "Its description is: \n",
      "lcl|KC296460.1_cds_AGH58093.1_1 [gene=HLA-A] [protein=HLA-A] [frame=3] [protein_id=AGH58093.1] [location=join(<1..269,511..>787)] [gbkey=CDS]\n",
      "Its sequence contains 544 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|KC296459.1_cds_AGH58092.1_1:\n",
      "Its description is: \n",
      "lcl|KC296459.1_cds_AGH58092.1_1 [gene=HLA-A] [protein=HLA-A] [frame=3] [protein_id=AGH58092.1] [location=join(<1..269,511..>787)] [gbkey=CDS]\n",
      "Its sequence contains 544 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|KC296458.1_cds_AGH58091.1_1:\n",
      "Its description is: \n",
      "lcl|KC296458.1_cds_AGH58091.1_1 [gene=HLA-A] [protein=HLA-A] [frame=3] [protein_id=AGH58091.1] [location=join(<1..269,511..>787)] [gbkey=CDS]\n",
      "Its sequence contains 544 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|KC296457.1_cds_AGH58090.1_1:\n",
      "Its description is: \n",
      "lcl|KC296457.1_cds_AGH58090.1_1 [gene=HLA-A] [protein=HLA-A] [frame=3] [protein_id=AGH58090.1] [location=join(<1..269,511..>787)] [gbkey=CDS]\n",
      "Its sequence contains 544 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|KC296456.1_cds_AGH58089.1_1:\n",
      "Its description is: \n",
      "lcl|KC296456.1_cds_AGH58089.1_1 [gene=HLA-A] [protein=HLA-A] [frame=3] [protein_id=AGH58089.1] [location=join(<1..269,511..>787)] [gbkey=CDS]\n",
      "Its sequence contains 544 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|KC296447.1_cds_AGH58088.1_1:\n",
      "Its description is: \n",
      "lcl|KC296447.1_cds_AGH58088.1_1 [gene=HLA-A] [protein=HLA-A] [frame=3] [protein_id=AGH58088.1] [location=join(<1..269,511..>787)] [gbkey=CDS]\n",
      "Its sequence contains 544 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|M64255.1_cds_AAA59610.1_1:\n",
      "Its description is: \n",
      "lcl|M64255.1_cds_AAA59610.1_1 [gene=HLA-A] [db_xref=GDB:G00-119-310] [protein_id=AAA59610.1] [location=1..147] [gbkey=CDS]\n",
      "Its sequence contains 147 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|M64253.1_cds_AAA59609.1_1:\n",
      "Its description is: \n",
      "lcl|M64253.1_cds_AAA59609.1_1 [gene=HLA-A] [db_xref=GDB:G00-119-310] [protein_id=AAA59609.1] [location=1..243] [gbkey=CDS]\n",
      "Its sequence contains 243 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|M64259.1_cds_AAA59608.1_1:\n",
      "Its description is: \n",
      "lcl|M64259.1_cds_AAA59608.1_1 [gene=HLA-A] [db_xref=GDB:G00-119-310] [protein_id=AAA59608.1] [location=1..147] [gbkey=CDS]\n",
      "Its sequence contains 147 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|M64257.1_cds_AAA59607.1_1:\n",
      "Its description is: \n",
      "lcl|M64257.1_cds_AAA59607.1_1 [gene=HLA-A] [db_xref=GDB:G00-119-310] [protein_id=AAA59607.1] [location=1..244] [gbkey=CDS]\n",
      "Its sequence contains 244 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|X94571.1_cds_CAA64263.1_1:\n",
      "Its description is: \n",
      "lcl|X94571.1_cds_CAA64263.1_1 [gene=HLA-A] [db_xref=HGNC:HGNC:4931,UniProtKB/TrEMBL:Q95360] [frame=3] [protein_id=CAA64263.1] [location=<1..>550] [gbkey=CDS]\n",
      "Its sequence contains 548 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|BA000025.2_cds_BAB63292.1_1:\n",
      "Its description is: \n",
      "lcl|BA000025.2_cds_BAB63292.1_1 [gene=C2] [partial=3'] [protein_id=BAB63292.1] [location=complement(join(<56..221,338..449,546..640,766..906,4546..4635,6407..6547,7803..7941,9565..9698,9899..9997,10081..10254,14896..15081,15649..15858,16015..16060))] [gbkey=CDS]\n",
      "Its sequence contains 1733 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|BA000025.2_cds_BAB63293.1_2:\n",
      "Its description is: \n",
      "lcl|BA000025.2_cds_BAB63293.1_2 [gene=NG35] [protein_id=BAB63293.1] [location=42521..43900] [gbkey=CDS]\n",
      "Its sequence contains 1380 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|BA000025.2_cds_BAB63294.1_3:\n",
      "Its description is: \n",
      "lcl|BA000025.2_cds_BAB63294.1_3 [gene=NG36] [protein_id=BAB63294.1] [location=join(46858..46906,47005..47223,47313..47566,50170..50202)] [gbkey=CDS]\n",
      "Its sequence contains 555 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|BA000025.2_cds_BAB63295.1_4:\n",
      "Its description is: \n",
      "lcl|BA000025.2_cds_BAB63295.1_4 [gene=G9A] [protein_id=BAB63295.1] [location=join(50967..51006,51098..51138,51270..51425,54234..54368,54494..54608,54766..54867,55089..55237,55330..55473,55560..55795,55875..56057,56177..56259,56735..56841,56939..57061,57154..57255,58820..58924,59037..59191,59277..59444,59886..60030,60397..60474,60827..60942,61051..61137,62744..62822,62988..63163,63572..63752)] [gbkey=CDS]\n",
      "Its sequence contains 3006 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|BA000025.2_cds_BAB63296.1_5:\n",
      "Its description is: \n",
      "lcl|BA000025.2_cds_BAB63296.1_5 [gene=NG22] [protein_id=BAB63296.1] [location=join(64856..64895,66754..66802,67707..67780,67906..67984,68808..68907,68990..69115,69316..69376,72276..72363,72440..72523,72790..73025,73135..73234,74198..74324,74606..74680,77705..77958,78047..78141,78230..78333,78443..78516,78751..78821,78921..79015,79095..79179,80083..80204)] [gbkey=CDS]\n",
      "Its sequence contains 2139 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|BA000025.2_cds_BAB63297.1_6:\n",
      "Its description is: \n",
      "lcl|BA000025.2_cds_BAB63297.1_6 [gene=NEU1] [protein=lysosomal sialidase] [protein_id=BAB63297.1] [location=join(81055..81213,81640..81832,82381..82643,83210..83392,83567..83789,83886..84112)] [gbkey=CDS]\n",
      "Its sequence contains 1248 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|BA000025.2_cds_BAB63298.1_7:\n",
      "Its description is: \n",
      "lcl|BA000025.2_cds_BAB63298.1_7 [gene=G8] [protein_id=BAB63298.1] [location=complement(join(104348..104468,106575..106681))] [gbkey=CDS]\n",
      "Its sequence contains 228 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|BA000025.2_cds_BAB63299.1_8:\n",
      "Its description is: \n",
      "lcl|BA000025.2_cds_BAB63299.1_8 [gene=HSPA1B] [protein=heat shock protein] [protein_id=BAB63299.1] [location=complement(114129..116054)] [gbkey=CDS]\n",
      "Its sequence contains 1926 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|BA000025.2_cds_BAB63300.1_9:\n",
      "Its description is: \n",
      "lcl|BA000025.2_cds_BAB63300.1_9 [gene=HSPA1A] [protein=heat shock protein] [protein_id=BAB63300.1] [location=complement(126326..128251)] [gbkey=CDS]\n",
      "Its sequence contains 1926 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|BA000025.2_cds_BAB63301.1_10:\n",
      "Its description is: \n",
      "lcl|BA000025.2_cds_BAB63301.1_10 [gene=HSPA1L] [protein=heat shock protein] [protein_id=BAB63301.1] [location=132036..133961] [gbkey=CDS]\n",
      "Its sequence contains 1926 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|BA000025.2_cds_BAB63302.1_11:\n",
      "Its description is: \n",
      "lcl|BA000025.2_cds_BAB63302.1_11 [gene=SMRNP] [protein=small nuclear ribonuclear protein D homolog] [protein_id=BAB63302.1] [location=join(137261..137263,137876..137943,145514..145544,145972..146031,146139..146264)] [gbkey=CDS]\n",
      "Its sequence contains 288 nucleotides.\n",
      "----------------------------------------------------------\n",
      "Processing the record lcl|BA000025.2_cds_BAB63303.1_12:\n",
      "Its description is: \n",
      "lcl|BA000025.2_cds_BAB63303.1_12 [gene=VARS2] [protein=valyl tRNA synthetase] [protein_id=BAB63303.1] [location=join(148803..149189,150900..151034,151125..151263,151504..151628,151719..151803,152049..152149,152283..152410,158284..158448,158690..158771,159303..159422,159515..159623,159709..159803,160832..160922,161173..161297,161385..161488,161575..161733,161841..161931,162066..162171,162257..162327,162412..162537,162895..162999,163080..163148,163235..163313,163450..163577,163865..164020,164204..164410,164482..164593,164726..165051,166356..166424)] [gbkey=CDS]\n",
      "Its sequence contains 3795 nucleotides.\n"
     ]
    }
   ],
   "source": [
    "#-- GeneratorLen(seqs, 5)\n",
    "count = 0; maximum=30\n",
    "\n",
    "for record in mseq1.seq_records:\n",
    "    count += 1\n",
    "    if count == maximum: break\n",
    "\n",
    "    # Extract individual parts of the FASTA record\n",
    "    identifier = record.id\n",
    "    description = record.description\n",
    "    sequence = record.seq\n",
    "\n",
    "    # Example: adapt to extract features you are interested in\n",
    "    print('----------------------------------------------------------')\n",
    "    print('Processing the record {}:'.format(identifier))\n",
    "    print('Its description is: \\n{}'.format(description))\n",
    "    amount_of_nucleotides = len(sequence)\n",
    "    print('Its sequence contains {} nucleotides.'.format(amount_of_nucleotides))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter term and align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9158 records saved at fasta/HLA_A_only.fasta\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_records = mseq1.search_term(\"HLA-A\")\n",
    "\n",
    "typeSeq = \"A\"\n",
    "fname = \"%s%s_%s_only.fasta\"%(mseq1.root, mseq1.prjName, typeSeq)\n",
    "\n",
    "mseq1.write_seqs(seq_records, fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not find muscle at '%s' /media/flalix/5c1ba0b4-f897-451c-9068-ac5e57194590/flalix/seaview/muscle\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "typeSeq = \"A\"\n",
    "\n",
    "in_file  = \"%s%s_%s_only.fasta\"%(root, prjName, typeSeq)\n",
    "out_file = \"%s%s_%s_only_aligned.fasta\"%(root, prjName, typeSeq)\n",
    "\n",
    "mseq1.allign_with_muscle(in_file, out_file, muscleApp = r\"../../seaview/muscle\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cut and Fulfill"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mseq1.read_aligned(typeSeq = \"A\", pattern = None, numOfRecords=None, sufix = \"only\", showmessage=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
